### ISSUE TITLE
Parent models not validating children as expected

### ISSUE THREAD

**Issue by kjdchapman** [rails/rails#34696]
### Overview
I'm seeing behaviour from ActiveModels that I don't understand. My understanding is trying to `save` a model will:
- validate the model itself
- validate any related children
- fail to save if either of these validations fail.

However, I am able to save a model that has children that _should_ fail validation.

### Steps to reproduce
I've not been able to recreate this problem in isolation from our app... but it's open source, so you can grab it and see for yourself.

I've got a PR open here - https://github.com/alphagov/govwifi-admin/pull/370. The relevant part is [the `validates_associated` attribute on our Location model](https://github.com/alphagov/govwifi-admin/pull/370/files#diff-639fc2bd424698d6c788942dd1a1784f) - switching `validates_associated` on and off will cause one of two test groups to fail.

### Expected behaviour
Either `validates_associated` is not necessary, or including it does not cause a timeout.

Note that the timeout only happens if there are already many related models. I suspect it's an inefficient uniqueness-checking query rather than an infinite loop (as some of the docs on `validates_associated` warn about).

### Actual behaviour
Including `validates_associated` causes the parent model to hang on `save`.
Excluding `validates_associated` allows a parent model to be saved with invalid children.

### System configuration
**Rails version**:
5.2.2

**Ruby version**:
2.5.3

**Comment by gmcgibbon** [comment]
Using this [test script](https://gist.github.com/gmcgibbon/b2d7a4230006d00187e35b6ebd73fcda) I was able to deduce that `validates_associated` on a parent and `accepts_nested_attributes_for` on the child's `belongs_to` relation seem to not play well. This behaviour seems to explain the timeout failure on your ip spec.

I don't understand why your location spec would fail without `validates_associated`. When you create nested active record objects, they are validated when saved or updated.

I'll try to dig deeper on the first point, it seems like a bug to me.

**Comment by kjdchapman** [comment]
Thanks for looking into this, @gmcgibbon.

> I don't understand why your location spec would fail without validates_associated. When you create nested active record objects, they are validated when saved or updated.

This is really the key problem for me. I don't have any need to include `validates_associated :ips`, that was just an attempt to solve my real need - getting IPs correctly validated. 

> I don't understand why your location spec would fail without validates_associated. When you create nested active record objects, they are validated when saved or updated.

Agreed - the tests on the pull request I linked seem to challenge this is working as intended. [Here's the specific block of tests that fail](https://github.com/alphagov/govwifi-admin/pull/370/files#diff-028488023c0543f941e51495a060fd50R42).

**Comment by kjdchapman** [comment]
We've resolved part of this issue at our end.

I'd missed that we have `accepts_nested_attributes_for` on **both** the parent and child model, which was causing the timeout issue (as @gmcgibbon described, and I failed to understand!). 

After removing this from the child model (`Ip`) we are now able to add `validates_associated` to get our desired behaviour.

The same quote as above is still outstanding as a point of confusion for me:
> I don't understand why your location spec would fail without validates_associated. When you create nested active record objects, they are validated when saved or updated.

I'm happy to close this issue (as my problem is resolved) or keep it open (as there's surprising behaviour I don't understand, but can replicate, albeit not in isolation).
